R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,530 @@
+#' influenced by others within a call, and which are influencers.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#' @return A `list` (of length of input `calls_pd`) where each element represents one call. Each element consists of a
+#' character vector containing names of objects that were influenced by this call, and names of objects influencing this
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`.
+#' @keywords internal
+#' @noRd
+extract_occurrence <- function(calls_pd) {
+
+  used_in_function <- function(call, rows) {
+
+    # If an object is a function parameter,
+    # then in calls_pd there is a `SYMBOL_FORMALS` entry for that object.
+    objects <- sapply(rows, function(row) call$text[row])
+
+    vapply(objects, function(object) {
+      if (any(call[call$token == "SYMBOL_FORMALS", "text"] == object) && any(call$token == "FUNCTION")) {
+        object_sf_ids <- call[call$text == object & call$token == "SYMBOL", "id"]
+        function_start_id <- call[call$token == "FUNCTION", "id"]
+        all(object_sf_ids > function_start_id)
+      } else {
+        FALSE
+      }
+    },
+      logical(1)
+    )
+  }
+
+  extract_functions <- function(calls_pd) {
+    # Detect functions created within code.
+    pd <- do.call(rbind, calls_pd)
+    symbols <- pd[pd$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"), c('token', 'text')]
+    symbols_table <- table(unique(symbols)$text)
+    names(symbols_table[symbols_table == 2])
+  }
+
+  functions <- extract_functions(calls_pd)
+
+  lapply(
+    calls_pd,
+    function(x) {
+      sym_cond <- which(x$token == "SYMBOL" | (x$token == "SYMBOL_FUNCTION_CALL" & x$text %in% functions))
+      sym_cond <- sym_cond[!used_in_function(x, sym_cond)]
+
+      # Watch out for SYMBOLS after $ and @. For x$a x@a: x is object, a is not.
+      # For x$a, a's ID is $'s ID-2 so we need to remove all IDs that have ID = $ID - 2.
+      dollar_ids <- x[x$token %in% c("'$'", "'@'"), "id"]
+      if (length(dollar_ids)) {
+        object_ids <- x[sym_cond, "id"]
+        after_dollar <- object_ids[(object_ids - 2) %in% dollar_ids]
+        sym_cond <- setdiff(sym_cond, which(x$id %in% after_dollar))
+      }
+
+      # If there was an assignment operation detect direction of it.
+      if (length(sym_cond) >= 2) {
+        ass_cond <- grep("ASSIGN", x$token)
+        if (length(ass_cond)) { # NOTE(1)
+          sym_cond <- sym_cond[sym_cond > ass_cond] # NOTE(2)
+          text <- x[sort(c(sym_cond, ass_cond)), "text"]
+
+          ans <- if (text[1] == "->") { # NOTE(3)
+            rev(text[-1])
+          } else {
+            text[-1]
+          }
+          c(ans[1], ":", unique(ans[-1]))
+        } else {
+          x[sym_cond, "text"]
+        }
```

Comment:
`text` and `ans` could be omitted and code could be simpler.

```suggestion
        if (length(ass_cond)) { # NOTE(1)
          sym_cond <- sym_cond[sym_cond > ass_cond] # NOTE(2)
          if (x$text[ass_cond] == "->") { # NOTE(3)
            sym_cond <- rev(sym_cond)
          } 
          c(x[sym_cond[1], "text"], ":", unique(x[-sym_cond[1], "text"]))
        } else {
          x[sym_cond, "text"]
        }
```

Reply:
needed to read it a couple of times! thanks, this is nice, will verify in a few minutes

Reply:
Cool, just reviewed against tests. This is great, thanks. Merging